COM-1357 Document the Stream usage in the Platform C# SDK

ClaudineL opened the pull request:
Copied from the documentation in Gino Vincelli's fork since he could not create a branch in this repo.

COM-1357 Document the Stream usage in the Platform C# SDK

ClaudineL commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+### Requirements:
+
+* Before you proceed with this method, ensure you’ve understood how to [index data](https://docs.coveo.com/en/2956/coveo-for-commerce/index-commerce-catalog-content-with-the-stream-api) and [create a commerce catalog](https://docs.coveo.com/en/3139/coveo-for-commerce/create-a-coveo-commerce-catalog).
```
I think the order is a bit weird, shouldn't the creation of the commerce catalog come first? I would also say 'index data with the Stream API'.

GinoVin replied:
> 

I apologize if this is causing extra work on your end. 
The next time around I'll be sure to ask for the proper access to the repo.

GinoVin commented:
Hi @ClaudineL 
Could you also please add Spuffynism @rtherien & @acarpentier01 as reviewers
ty!

GinoVin commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+### Requirements:
+
+* Before you proceed with this method, ensure you’ve understood how to [index data](https://docs.coveo.com/en/2956/coveo-for-commerce/index-commerce-catalog-content-with-the-stream-api) and [create a commerce catalog](https://docs.coveo.com/en/3139/coveo-for-commerce/create-a-coveo-commerce-catalog).
```
@SLangevin We had mentioned that there is a need to link to the docs for this information. Do you think it belongs here as something that needs to be understood to proceed?

GinoVin replied:
@SLangevin 👁️ 👁️

GinoVin commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+### Call Wrappers for the Stream API
+
+Call wrappers for the Stream API are found in the StreamApiDocumentServiceManager. You will need to create the [catalog source](https://docs.coveo.com/en/3295/index-content/add-or-edit-a-catalog-source) in the Admin UI first. Batching documents is recommended.
```
Should I link "Batching documents" to here for context?https://docs.coveo.com/en/90/index-content/manage-batches-of-items-in-a-push-source?

rtherien replied:
```suggestion
Call wrappers for the Stream API are found in the `StreamApiDocumentServiceManager`. You will need to create the [catalog source](https://docs.coveo.com/en/3295/index-content/add-or-edit-a-catalog-source) in the Coveo Administration Console first. Batching documents is recommended.
```

ClaudineL commented:
> Hi @ClaudineL Could you also please add Spuffynism @rtherien & @acarpentier01 as reviewers ty!

I wasn't able to add the last two, they don't show up when I type them up.

GinoVin commented:
> > Hi @ClaudineL Could you also please add Spuffynism @rtherien & @acarpentier01 as reviewers ty!
> 
> I wasn't able to add the last two, they don't show up when I type them up.

I checked with them, and they were pinged to review, so should be okay :)

rtherien commented on README.md:
```diff
@@ -100,7 +100,7 @@ document.JsonObjectSize > Constants.COMPRESSED_DATA_MAX_SIZE_IN_BYTES
 
 ### Pushing a document with large binary data size
-The Push API has a hard limit of 5MB (compressed) on the document's binary data. When the size exceeds that value you typically need to request an upload URI to an S3 Bucket. You would then put your document binary data in that bucket and refer to it by ID when pushing the target document with the Push API. However, lucky you, the SDK handles this automatically. Thus, the SDK compresses the conte ...
+The Push API has a hard limit of 5MB (compressed) on the document's binary data. When the size exceeds that value you typically need to request an upload URI to an S3 Bucket. You would then put your document binary data in that bucket and refer to it by ID when pushing the target document with the Push API. However, lucky you, the SDK handles this automatically. Thus, the SDK compresses the conte ...
```
```suggestion
The Push API has a hard limit of 5MB (compressed) on the document's binary data. When the size exceeds that value, you typically need to request an upload URI to an S3 Bucket. You would then put your document binary data in that bucket and refer to it by ID when pushing the target document with the Push API. However, lucky you, the SDK handles this automatically. Thus, the SDK compresses the content, and if it exceeds 5MB (compressed), it will do the needed logic. Take note that the maximum size of a document is 256MB (compressed).
```

rtherien replied:
Here: "You would then put your document binary data in that bucket and refer to it by ID when pushing the target document with the Push API"

I suppose we refer to the bucket's ID? If so, I'd be a bit more specific. It's a bit confusing whether we are referring to the binary data's ID or the bucket's ID.

GinoVin replied:
@ClaudineL I think you made this change. Would you be able to confirm :)

ClaudineL replied:
The only change I made was a space at the end of the sentence that was deleted 🙂 But if you think the original doc is too confusing I'll clarify it.

ClaudineL replied:
Looking at the code, the ID actually refers to the fileId returned by the request to get a file container, so I'm not sure if referring to it as a bucket is accurate, it's a separate kind of data structure.

rtherien commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+* Before you proceed with this method, ensure you’ve understood how to [index data](https://docs.coveo.com/en/2956/coveo-for-commerce/index-commerce-catalog-content-with-the-stream-api) and [create a commerce catalog](https://docs.coveo.com/en/3139/coveo-for-commerce/create-a-coveo-commerce-catalog).
+* For additional information and payload examples, refer to [How to Stream Your Catalog Data to Your Source](https://docs.coveo.com/en/lb4a0344/coveo-for-commerce/how-to-stream-your-catalog-data-to-your-source).
+* To use the Stream API, a version of the CoveoPlatformConfig needs to be created where useStreamApi is set to true:
```
```suggestion
* To use the Stream API, a version of the `CoveoPlatformConfig` needs to be created where `useStreamApi` option is set to true:
```

ClaudineL replied:
I would say `parameter` rather than option, but noted.

rtherien commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+Call wrappers for the Stream API are found in the StreamApiDocumentServiceManager. You will need to create the [catalog source](https://docs.coveo.com/en/3295/index-content/add-or-edit-a-catalog-source) in the Admin UI first. Batching documents is recommended.
+
+As the class inherits from DocumentServiceManager, the only truly new public calls are:
```
```suggestion
As the class inherits from `DocumentServiceManager`, the only truly new public calls are:
```

rtherien commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+### How to Open a Stream:
+
+> :warning: When you open and close a stream all previous files not indexed in the current operation will be removed from the index. Updating individual documents should be done in **update mode**, without the need to open a stream and close it afterwards.
```
```suggestion
> :warning: When you open and close a stream, all previous files not indexed in the current operation will be removed from the index. Updating individual documents should be done in **update mode**, without the need to open a stream and close it afterwards.
```

rtherien commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+### Add or Update Documents:
+
+The below calls use the call to `AddOrUpdateDocuments` that is in the DocumentServiceManager, whether or not the streamApiDocumentServiceManager contains a `streamId` will determine if the documents are uploaded in **stream mode** using the /chunk endpoint, or uploaded in **update mode** using the /files endpoint.
```
Would that make sense?

```suggestion
The below calls use the call to the `AddOrUpdateDocuments` endpoint that is in the` DocumentServiceManager`. The presence of a `streamId` in the `streamApiDocumentServiceManager` will determine if the documents are uploaded in **stream mode** using the /chunk endpoint, or uploaded in **update mode** using the `/files` endpoint.
```

rtherien commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+For multiple files, use: `streamApiDocumentServiceManager.AddOrUpdateDocuments(sourceId, documentsToAddOrUpdate, orderingId)`
+
+There shouldn’t be a need to manually call `GetNewChunkForStream`, as in **stream mode** each new call to `AddOrUpdateDocuments` will also call `GetNewChunkForStream`.
+In **update mode**, the call to /update is done automatically after uploading each batch.
```
```suggestion
In **update mode**, the call to the `/update` endpoint is done automatically after uploading each batch.
```

rtherien commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+OR
+To delete multiple files, use: `DeleteDocuments(sourceId, documentsToDelete, orderingId)`
+
+These calls will close an existing stream and call `AddOrUpdateDocuments` as well, followed by a call to the /update endpoint.
```
```suggestion
These calls will close an existing stream and call `AddOrUpdateDocuments` as well, followed by a call to the `/update` endpoint.
```

rtherien reviewed (approved):
Made a few suggestions, please verify that they're all accurate before applying.

GinoVin commented on README.md:
```diff
@@ -163,6 +163,67 @@ IList<string> documentsToDelete = new List<string> {
+
+### Requirements:
+
+* Before you proceed with this method, ensure you’ve understood how to [index data](https://docs.coveo.com/en/2956/coveo-for-commerce/index-commerce-catalog-content-with-the-stream-api) and [create a commerce catalog](https://docs.coveo.com/en/3139/coveo-for-commerce/create-a-coveo-commerce-catalog).
```
```suggestion
* Before you proceed with this method, ensure you’ve understood how to [index data with the stream API](https://docs.coveo.com/en/2956/coveo-for-commerce/index-commerce-catalog-content-with-the-stream-api) and [create a commerce catalog](https://docs.coveo.com/en/3139/coveo-for-commerce/create-a-coveo-commerce-catalog).
```

ClaudineL commented:
> @ClaudineL Hi again, Looks like I'm unable to commit any of the suggestions left here, would you be able to complete the commits (if they are valid) :)

Some of the suggestions are incorrect so I wouldn't accept all of them as-is, I'll go through each of them separately and update the document.

ClaudineL commented:
@Spuffynism Are you sure this is the right link? The section above it seems more topical: https://docs.coveo.com/en/lb4a0344/coveo-for-commerce/how-to-stream-your-catalog-data-to-your-source#step-2-upload-your-catalog-data-into-the-stream.
For batching they'd just need to call `AddOrUpdateDocuments` for each batch of ~256 MB.

ClaudineL commented:
I moved things around, clarified some things and changed the formatting to match more closely the rest of the README. @rtherien , @Spuffynism , @GinoVin and @SLangevin Did you wanna take another look?

GinoVin commented:
@ClaudineL Looks good!
One more review is inbound from @acarpentier01

acarpentier01 reviewed (commented):
Looks good! Just a few minor edits & two general comments:
1. Please use **see** <doc page> references instead of **visit** <doc page> references
2. Ensure headings are consistent (either use a gerund (e.g., Deleting xyz) or an imperative verb (e.g., Delete xyz), but not a mix of both)

GinoVin reviewed (approved):
I think I'm okay with these changes. 
Let me know when you merge the branch as I also have a small change to publish in the public documentation :)
